source/adios2/toolkit/burstbuffer/FileDrainer.h

```diff
@@ -0,0 +1,149 @@
+/*
+ * Distributed under the OSI-approved Apache License, Version 2.0.  See
+ * accompanying file Copyright.txt for details.
+ *
+ * FileDrainer.h
+ *
+ *  Created on: April 1, 2020
+ *      Author: Norbert Podhorszki <pnorbert@ornl.gov>
+ */
+
+#ifndef ADIOS2_TOOLKIT_BURSTBUFFER_FILEDRAINER_H_
+#define ADIOS2_TOOLKIT_BURSTBUFFER_FILEDRAINER_H_
+
+#include <fstream>
+#include <iostream>
+#include <locale>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <streambuf>
+#include <string>
+
+#include "adios2/common/ADIOSTypes.h"
+
+namespace adios2
+{
+namespace burstbuffer
+{
+
+enum class DrainOperation
+{
+    SeekEnd, // Seek to the end of target file toFileName (for future
+             // copyAppend). Seeking to End of fromFile is not allowed
+             // since another thread is writing to it
+    CopyAt,  // DO NOT USE: Copy countBytes from fromOffset to toOffset (does
+             // seek)
+    Copy,    // Copy countBytes (without seek)
+    WriteAt, // Write data from memory to toFileName directly at offset
+    Write,   // Write data from memory to toFileName directly (without seek)
+    Create,  // Open file for writing (creat) - only toFile
+    Open     // Open file for append - only toFile
+
+};
+
+struct FileDrainOperation
+{
+    DrainOperation op;
+    std::string fromFileName;
+    std::string toFileName;
+    size_t countBytes;
+    size_t fromOffset;
+    size_t toOffset;
+    std::vector<char> dataToWrite; // memory to write with Write operation
+
+    FileDrainOperation(DrainOperation op, const std::string &fromFileName,
+                       const std::string &toFileName, size_t countBytes,
+                       size_t fromOffset, size_t toOffset, const void *data);
+};
+
+typedef std::map<std::string, std::shared_ptr<std::ifstream>> InputFileMap;
```

Comment:
Not sure if `std::shared_ptr` is needed at all. `ifstream` and `ofstream` can exist (as `false`) without being opened. I'd stick with objects as possible to prevent managing ownership.

Reply:
I started with that first. Then the XL compiler complained that I was trying to use a deleted copy constructor (of ofstream). I did not know how to solve it, so moved to using shared ptr. 

I could not figure out how to emplace a new ofstream object into the map and then return a reference back to the caller that xlc++ would accept.

Reply:
A comment would be helpful since it's very valuable info on the XL compilers (we'll probably forget). Also, `fstream` doesn't have a copy constructor (only move), so maybe `std::unique_ptr` is more suitable?

Reply:
Unless this is accessed concurrently, then nevermind since we can't pass `std::unique_ptr &` around. I'd keep `std::shared_ptr` and pass `std::shared_ptr&` as much as possible.

Reply:
If you pass the `std::shared_ptr` by reference, the reference count is not incremented. Is that what is intended?